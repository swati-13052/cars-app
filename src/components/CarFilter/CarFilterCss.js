import styled from "styled-components";
import InputLabel from '@mui/material/InputLabel';
import Button from '@mui/material/Button';

export const CarsFilterContainer = styled.div`
    width: 25%;
    height: 230px;
    border: 1px solid;
    border-color: ${(props) => props.theme.COLOURS.GRAY_LIGHT};
    padding: 24px;

    .filter-button-container {
        display: flex;
        justify-content: end;
    }
`

export const StyledInputLabel = styled(InputLabel)`
&&&{
    margin-bottom: 8px;
    font-size: 14px;
}
`

export const FilterButton = styled(Button)`
    &&& {
        margin-top: 16px;
        text-align: right;
        width:40%;

    }
`
