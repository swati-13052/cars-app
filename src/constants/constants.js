export const CAR_FILTERS = {
    color: 'color',
    manufacturer: 'manufacturer'
}