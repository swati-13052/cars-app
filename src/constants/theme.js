import { createTheme } from "@mui/material";
import { createContext } from "react";

export const theme = {
    COLOURS: {
        GRAY_LIGHT: '#EDEDED',
        GRAY_DARK: '#4A4A4A',
        ORANGE_LIGHT: '#EA7F28',
        ORANGE_DARK: '#D37324'
    },
    SPACING: {
        SMALL: '8px',
        MEDIUM: '12px',
        LARGE: '32px',
    }
}

export const ThemeContext = createContext();

export const ButtonTheme = createTheme({
    palette: {
        primary: {
            light: '#ff7961',
            main: '#EA7F28',
            dark: '#D37324',
            contrastText: '#fff'
        },
        action: {
            hover: '#D37324'
        }
    }
})