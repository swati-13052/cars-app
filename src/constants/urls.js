import { payloadToQueryParams } from './../utils/util';

export const getAllCarsURL = (queryParams) => `https://auto1-mock-server.herokuapp.com/api/cars${payloadToQueryParams(queryParams)}`;
export const getAllColoursURL = `https://auto1-mock-server.herokuapp.com/api/colors`;
export const getAllManufacturerURL = `https://auto1-mock-server.herokuapp.com/api/manufacturers`;
export const getCarDetails = (stockNumber) => `https://auto1-mock-server.herokuapp.com/api/cars/${stockNumber}`;
export const logoURL = `https://auto1-js-task-api--mufasa71.repl.co/images/logo.png`;