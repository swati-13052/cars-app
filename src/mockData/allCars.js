export const allCars = [
    {
        "stockNumber": 30257,
        "manufacturerName": "Skoda",
        "modelName": "Forman",
        "color": "green",
        "mileage": {
            "number": 100099,
            "unit": "km"
        },
        "fuelType": "Petrol",
        "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    }

]