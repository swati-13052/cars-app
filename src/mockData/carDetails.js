export const carDetails = {
    "car": {
      "stockNumber": 10035,
      "manufacturerName": "Mercedes-Benz",
      "modelName": "GL-Klasse",
      "color": "blue",
      "mileage": {
        "number": 180222,
        "unit": "km"
      },
      "fuelType": "Diesel",
      "pictureUrl": "https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
    }
  }