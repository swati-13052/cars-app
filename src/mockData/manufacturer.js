export const manufacturers = [
    {
        "name": "Audi",
        "models": [
            {
                "name": "100"
            },
            {
                "name": "200"
            },
            {
                "name": "50"
            },
            {
                "name": "80/90"
            },
            {
                "name": "A1"
            },
            {
                "name": "A2"
            },
            {
                "name": "A3"
            },
            {
                "name": "A4"
            },
            {
                "name": "A4 Allroad"
            },
            {
                "name": "A5"
            },
            {
                "name": "A6"
            },
            {
                "name": "A6 Allroad"
            },
            {
                "name": "A7"
            },
            {
                "name": "A8"
            },
            {
                "name": "Allroad"
            },
            {
                "name": "Cabriolet"
            },
            {
                "name": "Coupe"
            },
            {
                "name": "Q2"
            },
            {
                "name": "Q3"
            },
            {
                "name": "Q5"
            },
            {
                "name": "Q7"
            },
            {
                "name": "Quattro"
            },
            {
                "name": "R8"
            },
            {
                "name": "RS 3"
            },
            {
                "name": "RS 4"
            },
            {
                "name": "RS 5"
            },
            {
                "name": "RS 6"
            },
            {
                "name": "RS 7"
            },
            {
                "name": "RS Q3"
            },
            {
                "name": "S1"
            },
            {
                "name": "S3"
            },
            {
                "name": "S4"
            },
            {
                "name": "S5"
            },
            {
                "name": "S6"
            },
            {
                "name": "S7"
            },
            {
                "name": "S8"
            },
            {
                "name": "SQ5"
            },
            {
                "name": "SQ7"
            },
            {
                "name": "TT"
            },
            {
                "name": "TT Roadster"
            },
            {
                "name": "V8"
            }
        ]
    }
]