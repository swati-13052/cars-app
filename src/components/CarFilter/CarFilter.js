import MenuItem from '@mui/material/MenuItem';
import Select from '@mui/material/Select';
import { useContext, useEffect, useState } from 'react';
import { CAR_FILTERS } from 'constants/constants';
import { ThemeContext } from 'constants/theme';
import { CarsFilterContainer, FilterButton, StyledInputLabel } from './CarFilterCss';
import { firstCharUpperCase } from 'utils/util';

function CarFilter(props) {
    const theme = useContext(ThemeContext);
    const [filters, setFilters] = useState({});

    const setFilter = (event, filterName) => {
        setFilters({ ...filters, [filterName]: event.target.value });
    }

    const handleFilterChange = () => {
        props.handleFilterChange(filters);
    }

    useEffect(() => {
        setFilters({ ...props.filters })
    }, [props.filters])

    return (
        <CarsFilterContainer theme={theme}>
            <StyledInputLabel data-testid="color-label" id="color-label">Color</StyledInputLabel>
            <Select
                labelId="color-label"
                value={filters?.color || ''}
                data-testid="color-select"
                onChange={(event) => setFilter(event, CAR_FILTERS.color)}
                displayEmpty
                sx={{
                    width: '100%',
                    marginBottom: '8px'
                }}
                inputProps={{ 'aria-label': 'Without label' }}
            >
                <MenuItem value="">All car colors</MenuItem>
                {props.allColours?.map((color, index) => {
                    return (
                        <MenuItem value={color} key={index}>{firstCharUpperCase(color)}</MenuItem>
                    )
                })}
            </Select>

            <StyledInputLabel data-testid="manufacturer-label" id="manufacturer-label">Manufacturer</StyledInputLabel>
            <Select
                labelId="manufacturer-label"
                value={filters?.manufacturer || ''}
                data-testid="manufacturer-select"
                onChange={(event) => setFilter(event, CAR_FILTERS.manufacturer)}
                displayEmpty
                sx={{
                    width: '100%'
                }}
                inputProps={{ 'aria-label': 'Without label' }}
            >
                <MenuItem value="">All Manufacturers</MenuItem>
                {props.allManufacturer?.map((manufacturer, index) => {
                    return (
                        <MenuItem value={manufacturer.name} key={index}>{manufacturer.name}</MenuItem>
                    )
                })}
            </Select>
            <div className="filter-button-container">
                <FilterButton onClick={() => handleFilterChange()} color="primary" pagetheme={theme} variant="contained">Filter</FilterButton>
            </div>
        </CarsFilterContainer>
    )
}

export default CarFilter;