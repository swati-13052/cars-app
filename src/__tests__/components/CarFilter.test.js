import React from 'react'
import { render, cleanup, screen, fireEvent, within } from '@testing-library/react'
import CarFilter from './../../components/CarFilter/CarFilter';
import { theme, ThemeContext } from './../../constants/theme';
import { colors } from '../../mockData/color';
import { manufacturers } from '../../mockData/manufacturer';

const renderWithContext = (
    component) => {
    return {
        ...render(
            <ThemeContext.Provider value={theme}>
                {component}
            </ThemeContext.Provider>)
    }
}

afterEach(cleanup)

it('should take a snapshot', () => {
    const { asFragment } = renderWithContext(<CarFilter />)
    expect(asFragment(<CarFilter />)).toMatchSnapshot()
});

it('should display correct tex for the labels', () => {
    const { getByTestId } = renderWithContext(<CarFilter />)
    expect(getByTestId('color-label')).toHaveTextContent('Color');
    expect(getByTestId('manufacturer-label')).toHaveTextContent('Manufacturer');
})

it('should display color dropdown and change value on clicking any option', async () => {
    const { getByTestId } = renderWithContext(<CarFilter allColours={colors}/>);
    const dropdown = getByTestId('color-select');
    const defaultOption = dropdown.children[0];
    expect(dropdown).toHaveTextContent(defaultOption.textContent);
    
    fireEvent.mouseDown(screen.getByRole('button', {name: 'Color'}));

    const listbox = within(screen.getByRole('listbox'));
    const listBoxItems = listbox.getAllByRole('option');
    expect(listBoxItems.length).toBe(8);

    fireEvent.click(listbox.getByText(/Red/i));
    expect(getByTestId('color-select')).toHaveTextContent(/Red/i); 
})

it('should display manufacturer dropdown and change value on clicking any option', async () => {
    const { getByTestId } = renderWithContext(<CarFilter allManufacturer={manufacturers}/>);
    const dropdown = getByTestId('manufacturer-select');
    const defaultOption = dropdown.children[0];
    expect(dropdown).toHaveTextContent(defaultOption.textContent);
    
    fireEvent.mouseDown(screen.getByRole('button', { name: 'Manufacturer'}));

    const listbox = within(screen.getByRole('listbox'));
    const listBoxItems = listbox.getAllByRole('option');
    expect(listBoxItems.length).toBe(2);

    fireEvent.click(listbox.getByText(/Audi/i));
    expect(getByTestId('manufacturer-select')).toHaveTextContent(/Audi/i); 
})


