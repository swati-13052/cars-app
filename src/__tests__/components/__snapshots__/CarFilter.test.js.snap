// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ goaJdA"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary sc-gsnTZi kjnUNk css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      data-testid="color-label"
      id="color-label"
    >
      Color
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary  css-1g4duqy-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      data-testid="color-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Without label"
        aria-labelledby="color-label"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="button"
        tabindex="0"
      >
        All car colors
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
    <label
      class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary sc-gsnTZi kjnUNk css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      data-testid="manufacturer-label"
      id="manufacturer-label"
    >
      Manufacturer
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary  css-bx5laj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      data-testid="manufacturer-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Without label"
        aria-labelledby="manufacturer-label"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="button"
        tabindex="0"
      >
        All Manufacturers
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
    <div
      class="filter-button-container"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root sc-dkzDqf ldQsVp css-sghohy-MuiButtonBase-root-MuiButton-root"
        pagetheme="[object Object]"
        tabindex="0"
        type="button"
      >
        Filter
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
